File: lua/test/comparison_entity_test.lua
-- Comparison entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("we-ultrarich_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("ComparisonEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Comparison(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = comparison_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "comparison." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set WEULTRARICH_TEST_COMPARISON_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local comparison_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.comparison")))
    local comparison_ref01_data = nil
    if #comparison_ref01_data_raw > 0 then
      comparison_ref01_data = helpers.to_map(comparison_ref01_data_raw[1][2])
    end

    -- LOAD
    local comparison_ref01_ent = client:Comparison(nil)
    local comparison_ref01_match_dt0 = {}
    local comparison_ref01_data_dt0_loaded, err = comparison_ref01_ent:load(comparison_ref01_match_dt0, nil)
    assert.is_nil(err)
    assert.is_not_nil(comparison_ref01_data_dt0_loaded)

  end)
end)

function comparison_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/comparison/ComparisonTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read comparison test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "comparison01", "comparison02", "comparison03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("WEULTRARICH_TEST_COMPARISON_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["WEULTRARICH_TEST_COMPARISON_ENTID"] = idmap,
    ["WEULTRARICH_TEST_LIVE"] = "FALSE",
    ["WEULTRARICH_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["WEULTRARICH_TEST_COMPARISON_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["WEULTRARICH_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["WEULTRARICH_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["WEULTRARICH_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/comparison_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/we-ultrarich-sdk/go"
	"github.com/voxgig-sdk/we-ultrarich-sdk/go/core"

	vs "github.com/voxgig-sdk/we-ultrarich-sdk/go/utility/struct"
)

func TestComparisonEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Comparison(nil)
		if ent == nil {
			t.Fatal("expected non-nil ComparisonEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := comparisonBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "comparison." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WE_ULTRARICH_TEST_COMPARISON_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		comparisonRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath(setup.data, "existing.comparison")))
		var comparisonRef01Data map[string]any
		if len(comparisonRef01DataRaw) > 0 {
			comparisonRef01Data = core.ToMapAny(comparisonRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = comparisonRef01Data

		// LOAD
		comparisonRef01Ent := client.Comparison(nil)
		comparisonRef01MatchDt0 := map[string]any{}
		comparisonRef01DataDt0Loaded, err := comparisonRef01Ent.Load(comparisonRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if comparisonRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func comparisonBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "comparison", "ComparisonTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read comparison test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse comparison test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"comparison01", "comparison02", "comparison03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WE_ULTRARICH_TEST_COMPARISON_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WE_ULTRARICH_TEST_COMPARISON_ENTID": idmap,
		"WE_ULTRARICH_TEST_LIVE":      "FALSE",
		"WE_ULTRARICH_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["WE_ULTRARICH_TEST_COMPARISON_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WE_ULTRARICH_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
			},
			extraOpts,
		})
		client = sdk.NewWeUltrarichSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WE_ULTRARICH_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WE_ULTRARICH_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
